### ISSUE TITLE
Implement Specialized `GroupColumn` for `Date`/`Time`/`Timestamp` types

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#13263]
### Is your feature request related to a problem or challenge?

In https://github.com/apache/datafusion/pull/12269 @jayzhan211 made significant improvements to how group values are stored in multi-column aggregations. 

Specifically for queries like

```sql
SELECT ... FROM ... GROUP BY col1, ... colN
```

The improvement relies on implementing specialized versions of [`GroupColumn`](https://github.com/apache/datafusion/blob/a6586ccfffb569461e357817529ecf6647c6d62e/datafusion/physical-plan/src/aggregates/group_values/group_column.rs#L53) for the types of `col1`, `colN` 

We have implemented the primitive types and Strings/StringViews now, but we have not implemented all types

This means queries like 

```sql
SELECT ... FROM ... GROUP BY int_col, date_col
```

Will fall back to the slower (but general) `GroupValuesRows`: https://github.com/apache/datafusion/blob/a6586ccfffb569461e357817529ecf6647c6d62e/datafusion/physical-plan/src/aggregates/group_values/row.rs#L40-L41

### Describe the solution you'd like

Implement `GroupColumn` for all primitive types.

You can see how to do this here:

https://github.com/apache/datafusion/blob/e4bd57918b1725f37f3f500c503c07b1c1bf90bf/datafusion/physical-plan/src/aggregates/group_values/mod.rs#L117-L121

and the make sure there are tests for each of those types in queries that group on multiple columns

### Additional context

Here is an example for how this was done for Strings: https://github.com/apache/datafusion/pull/12809

**Comment by buraksenn** [comment]
take

I plan to wrap up nth_value PR today thanks to @jcsherin's reviews, then start on this one if it is okay. Otherwise, I can drop this

**Comment by buraksenn** [comment]
take

**Comment by jonathanc-n** [comment]
Another thing, should this be done before or after #13262? I'm thinking after, just to reduce overall work.

**Comment by alamb** [comment]
> Another thing, should this be done before or after #13262? I'm thinking after, just to reduce overall work.

I agree after would be good -- I am hoping that https://github.com/apache/datafusion/issues/13262 can be quickly banged out / merged.

However, I think most of the work for this PR will be testing (the actual code is likely not all that much) so that might be good to get ready as well

**Comment by buraksenn** [comment]
> @buraksenn Will you be working on this? if not, I would like to try working on this.

I was waiting #13262 PR but from that PR I see that you've more knowledge on this topic so I don't want to block this :D. Feel free to work if you've time. I'm dropping this issue I can take a look if you cant find time

**Comment by buraksenn** [comment]
drop

**Comment by jonathanc-n** [comment]
take

**Comment by jonathanc-n** [comment]
@alamb Do you want all three to be implemented in different files?
